src/Microsoft.Dnx.Watcher.Core/DnxWatcher.cs

```diff
@@ -0,0 +1,214 @@
+{
+    public class DnxWatcher
+    {
+        private readonly Func<string, IFileWatcher> _fileWatcherFactory;
+        private readonly Func<IProcessWatcher> _processWatcherFactory;
+        private readonly IProjectProvider _projectProvider;
+        private readonly ILoggerFactory _loggerFactory;
+
+        private readonly ILogger _logger;
+
+        public DnxWatcher(
+            Func<string, IFileWatcher> fileWatcherFactory,
+            Func<IProcessWatcher> processWatcherFactory,
+            IProjectProvider projectProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _fileWatcherFactory = fileWatcherFactory;
+            _processWatcherFactory = processWatcherFactory;
+            _projectProvider = projectProvider;
+            _loggerFactory = loggerFactory;
+
+            _logger = _loggerFactory.CreateLogger(nameof(DnxWatcher));
+        }
+        public async Task<bool> WatchAsync(string projectFile, string[] dnxArguments, string workingDir, CancellationToken cancellationToken)
+        {
+            dnxArguments = new string[] { "--project", projectFile }
+                .Concat(dnxArguments)
+                .Select(arg =>
+                {
+                    // If the argument has spaces, make sure we quote it
+                    if (arg.Contains(" ") || arg.Contains("\t"))
+                    {
+                        return $"\"{arg}\"";
+                    }
+
+                    return arg;
+                })
+                .ToArray();
+
+            var dnxArgumentsAsString = string.Join(" ", dnxArguments);
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var project = await WaitForValidProjectJsonAsync(projectFile, cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return true;
+                }
+
+                using (var currentRunCancellationSource = new CancellationTokenSource())
+                using (var combinedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
+                    cancellationToken,
+                    currentRunCancellationSource.Token))
+                {
+                    var fileWatchingTask = WaitForProjectFileToChangeAsync(project, combinedCancellationSource.Token);
+                    var dnxTask = WaitForDnxToExitAsync(dnxArgumentsAsString, workingDir, combinedCancellationSource.Token);
+
+                    var tasksToWait = new Task[] { dnxTask, fileWatchingTask };
+
+                    int finishedTaskIndex = Task.WaitAny(tasksToWait, cancellationToken);
+
+                    // Regardless of the outcome, make sure everything is cancelled
+                    // and wait for dnx to exit. We don't want orphan processes
+                    currentRunCancellationSource.Cancel();
+                    Task.WaitAll(tasksToWait);
+
+                    if (cancellationToken.IsCancellationRequested)
```

Comment:
If you want to return a cancelled task to the caller `cancellationToken.ThrowIfCancellationRequested()`